File: src/org/redrune/networking/packet/context/impl/GroundItemPickupPacketContext.java
package org.redrune.networking.packet.context.impl;

import org.redrune.game.entity.actor.player.Player;
import org.redrune.game.entity.actor.player.data.RouteEvent;
import org.redrune.game.entity.item.FloorItem;
import org.redrune.game.global.WorldTile;
import org.redrune.game.global.map.region.RegionManager;
import org.redrune.networking.packet.context.PacketContext;

/**
 * @author Tyluur <itstyluur@gmail.com>
 * @since 2019-02-04
 */
public class GroundItemPickupPacketContext extends PacketContext {
	
	private final int regionId;
	
	private final boolean forceRun;
	
	private final FloorItem item;
	
	private final WorldTile tile;
	
	public GroundItemPickupPacketContext(int regionId, boolean forceRun, FloorItem item, WorldTile tile) {
		this.regionId = regionId;
		this.forceRun = forceRun;
		this.item = item;
		this.tile = tile;
	}
	
	@Override
	public void handle(Player player) {
		if (forceRun) {
			player.setRunModeOn(true);
		}
		player.stopAll(false);
		player.setRouteEvent(new RouteEvent(item, () -> {
			final FloorItem item1 = RegionManager.getRegion(regionId).getGroundItem(item.getId(), tile, player);
			if (item1 == null) {
				return;
			}
			player.setNextFaceWorldTile(tile);
			RegionManager.removeGroundItem(player, item1);
		}, true));
	}
}

File: src/main/kotlin/org/redrune/networking/packet/context/impl/GroundItemPickupPacketContext.kt
package org.redrune.networking.packet.context.impl

import org.redrune.game.content.entity.actor.player.event.item.ItemFloorPickupEvent
import org.redrune.game.entity.actor.player.Player
import org.redrune.game.entity.item.FloorItem
import org.redrune.game.global.WorldTile
import org.redrune.networking.packet.context.PacketContext

/**
 * @author Tyluur <itstyluur@icloud.com>
 * @since 2019-02-04
 */
class GroundItemPickupPacketContext(
    private val regionId: Int,
    private val forceRun: Boolean,
    private val item: FloorItem,
    private val tile: WorldTile,
) : PacketContext() {
    override fun handle(player: Player) {
        player.eventManager.start(ItemFloorPickupEvent(regionId, forceRun, item, tile))
    }
}